### ISSUE TITLE
Benchmarking / Performance issue for Legacy vs. Class Based API

### ISSUE THREAD

**Issue by danielvdao** [rmosolgo/graphql-ruby#3490]
**Describe the bug**

I'm having trouble replicating the proper benchmarks / performance gains seen in the 1.9.21 (and latest version) of class-based API of graphql-ruby in reference to https://github.com/rmosolgo/graphql-ruby/issues/861. I'm trying to isolate the behavior in benchmarking code and would love some eyes.

Here is an attached gist with a few things: https://gist.github.com/danielvdao/b20a3cae4fa726728d6abd138fde3cfd

Legacy based:
- [Code used to run benchmark for legacy](https://gist.github.com/danielvdao/b20a3cae4fa726728d6abd138fde3cfd#file-legacy_based_benchmark-rb)
- [Benchmarks for legacy based API on 1.9.21](https://gist.github.com/danielvdao/b20a3cae4fa726728d6abd138fde3cfd#file-legacy-based-api-benchmarks)
- [Memory Profiling done for legacy on 1.9.21](https://gist.github.com/danielvdao/b20a3cae4fa726728d6abd138fde3cfd#file-legacy_based_graphql_1-9-21)
- [Memory Profiling done for legacy on 1.6.6
](https://gist.github.com/danielvdao/b20a3cae4fa726728d6abd138fde3cfd#file-legacy_based_memory_profile_1-6-6-txt)

Class based:
- [Code used to run benchmark for class based](https://gist.github.com/danielvdao/b20a3cae4fa726728d6abd138fde3cfd#file-class_based_api_benchmark-rb)
- [Class based API benchmarks](https://gist.github.com/danielvdao/b20a3cae4fa726728d6abd138fde3cfd#file-class-based-api-benchmarks)
- [Memory profiling done for class based on 1.9.21](https://gist.github.com/danielvdao/b20a3cae4fa726728d6abd138fde3cfd#file-class_based_memory_profile-txt)
- [Ruby profiling code for execution](https://gist.github.com/danielvdao/b20a3cae4fa726728d6abd138fde3cfd#file-ruby-profile-class-based)

It looks like memory profile is showing that the new interpreter is being used. It also shows that memory allocation is a lot less too (which seems to be expected) as `FieldResolutionContext` is 🔪  in the new runtime interpreter vs. legacy based behavior on 1.6.6.

That said... I'm not able to see a performance boost on `benchmark-ips` and the numbers look worse(?) on 1.9.21 (also seeing the same results on latest as well). I'm thinking either my code is wrong or something in my setup is not correct to match the same performance gains in https://github.com/rmosolgo/graphql-ruby/issues/861.

Would love your eyes on this @rmosolgo as I think this issue can also be a comprehensive summary + isolated benchmark of what gains one can make switching over to the new class-based API! ❤️ 

As a random plug I've also written a Rubocop cop to help catch legacy based DSL, so maybe these two in conjunction could help motivate people? 🙏🏽  https://github.com/DmitryTsepelev/rubocop-graphql/pull/39

**Versions**
`graphql` version: 1.9.21
System: Mac OS X local development

**Comment by rmosolgo** [comment]
Hey, thanks for the great writeup and details here. It looks to me like most of the time is spent in runtime bookkeeping around dead paths and non-null requirement tracking. Similar to https://github.com/rmosolgo/graphql-ruby/issues/3345, there's no easy wins I see here. 

As you noticed, the runtime footprint is reduced. For class-based schemas on 1.10+, the schema's memory footprint is greatly reduced, too. Those were my main targets. 

I think there might be some options for reducing the overhead: 

- Can `@dead_paths` be tracked another way? (Perhaps inside `HashResponse`?) 
- I think the recursive call to `Runtime#after_lazy` can be removed and replaced with `yield(...)`, that might reduce some overhead 
- I think the call to `set_interpreter_context` before `after_lazy` yields could be pushed up to the caller, but not called by default. removing it causes a test to fail, but I bet whatever codepath depends on it could just set context itself instead
- Maybe rearranging some of those `case` / `if/elsif/elsif/...` chains could improve performance by handling more common cases first. Likewise, maybe `when "NON_NULL"` could be handled as a special case at the top of the method 

In practice, the reason I haven't dug into these refactors in the past is because in my own application, it's I/O and application code that takes the lion's share of graphql execution time, not GraphQL overhead. (But as described in the issue linked above, some people _are_ impacted by graphql overhead!)

**Comment by eapache** [comment]
Related to `@dead_paths` and `HashResponse`, I have a few other observations and (definitely non-trivial) ideas for reducing overhead.
- Right now methods like `write_invalid_null_in_response` will walk the query from the root to the relevant field _4 full times_ by default (once in `dead_path?`, twice in `write_in_response` which also calls `dead_path?` then walks it again in `HashResponse`, and once in `add_dead_path`). For a deep query this can add up fast.
- Right now the interpreter tracks the full `path` from the root (which gets `dup`ed on every selection?!) and then writes into the response using that, thus always walking from the root on every field. Again this would add up really fast for a deep query.

It feels like there's a missing unified concept of a walkable "query/response" tree capable of tracking:
- the response value for each node
- dead paths
- other path/node-dependent metadata
- the interpreter's current location in the query/response

This would let us avoid all of this walking down from the root all the time? Unify `path`, `response`, and `@dead_paths` into a movable "you are here" pointer in this tree, and you could get rid of a lot of `dup`s and a lot of the looped hash-lookups.

**Comment by danielvdao** [comment]
@rmosolgo thank you for replying! I'll be honest and admit that I am a little out of my comfort zone here, so this is still unfamiliar territory to me. If I am wrong - please correct me.

Dead paths? Can you give a little bit more context here? Maybe a link in the code to help track its usage? Wondering if you used a tool to locate this and if I can try myself.

And non-null requirement tracking... do you mean the following? It sounds like the number of non-null types one has can cause a lot of memory allocation to add up vs. the legacy type.

```
allocated memory by class-----------------------------------
.....
      9368  GraphQL::NonNullType
```

A lot of our code is I/O bound as well. Our graphQL service is our API gateway, so most of the request time is bottled in the other services. As a result we try to mostly keep our graphQL layer thin to reduce application + business logic happening in that service.

**TL;DR ❤️ **
I am wondering how you were able to see a 50% performance increase in isolated benchmarks then? Were you running different versions? Referencing https://github.com/rmosolgo/graphql-ruby/issues/861#issuecomment-458533219. Was it a different set of code than the gist that was posted 🤔 ? I adapted the gist from there, so from a naïve standpoint I'd imagine that I'd just be able to replicate a similar set of numbers as you were.

I would be happy to at least try and spend some time digging into some of the code if you think there are easy cases of optimizations. The last one of `case` / `if/elsif/elseif/...` sounds not too difficult.

**Comment by danielvdao** [comment]
Thanks for updating some of the code @rmosolgo, I see you've pushed more changes to [https://github.com/rmosolgo/graphql-ruby/pull/3364/commits/3e3f42ecc2e2b7989a6896b7fd6b75e9d1201561](commit) to try and tackle this. For more debugging which you may have already done, I've added a profiler gist for the class based code above [here](https://gist.github.com/danielvdao/b20a3cae4fa726728d6abd138fde3cfd#file-ruby-profile-class-based).

**Comment by rmosolgo** [comment]
Sorry for the overwhelming list, @danielvdao. Mostly, I just posted those off-the-cuff ideas so I wouldn't forget them 😅 

@eapache , I started going down the same rabbit trail and it's mostly working: https://github.com/rmosolgo/graphql-ruby/pull/3494 Although there's still a few things to improve there, as noted.

**Comment by danielvdao** [comment]
Gotcha. No worries 🙂 . It looks like I have some learning I'll need to be doing too, so thanks for the list regardless! @rmosolgo to bug you again, but I was wondering if you happen to know the answer to this question:

> TL;DR ❤️
I am wondering how you were able to see a 50% performance increase in isolated benchmarks then? Were you running different versions? Referencing #861 (comment). Was it a different set of code than the gist that was posted 🤔 ? I adapted the gist from there, so from a naïve standpoint I'd imagine that I'd just be able to replicate a similar set of numbers as you were. - [issue comment](https://github.com/rmosolgo/graphql-ruby/issues/3490#issuecomment-846155892)

I would like to try and replicate the benchmark performance gains you saw, but was wondering if you have the code you were executing to run your benchmarks? I'm struggling to see if there's something different in my code.

**Comment by rmosolgo** [comment]
Oops, sorry I missed that before. The benchmark is linked from this comment: https://github.com/rmosolgo/graphql-ruby/issues/861#issuecomment-319418444

It's _very_ similar to the one you posted here. Interestingly, both results that we've posted are way faster than the ones in that old issue. I've gotten a new laptop, and we've had a few Ruby releases since then, so the difference might be due to those. Besides that, I'm not sure what version of GraphQL-Ruby I ran those benchmarks on. It's possible that changes in areas like schema definition, field extensions, or argument validations have changed the comparison.

**Comment by danielvdao** [comment]
@rmosolgo no worries. Thanks for the reply. I imagine it's hard to replicate at this point since it's been a few years(?) since the original issue so I understand there's already been so many different variables 😓 . And yep I saw that comment, I adapted my benchmark code from there as well so I was curious 🙂 .

That said, I'm excited by the work that's happening right now!! Thank you both immensely for being so helpful and responsive!

**Comment by rmosolgo** [comment]
How about giving it another go locally since #3494 is merged to master? You can bundle that version with 

```ruby 
gem "graphql", github: "rmosolgo/graphql-ruby"
```

Alternatively, I'll cut a new release here soon and you can try that ...

**Comment by rmosolgo** [comment]
🚢 'd in 1.12.11. If you get a chance to run your benchmarks and you're still seeing poor performance, please open a new issue with the results. Thanks for opening this -- what a fruitful investigation it turned out to be!

**Comment by danielvdao** [comment]
Thanks @rmosolgo. Yes, I'll make sure to tap back in once we migrate over. It's going to be a pretty large jump though, but I can provide some results once it's done 🙂 . Thanks so much for digging into this!
